File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/rtree/NodeList.java
package edu.uci.ics.jung.visualization.spatial.rtree;

import java.awt.geom.Rectangle2D;
import java.util.ArrayList;
import java.util.Collection;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A list of elements that are Bounded by a Rectangle2D The list is also Bounded by the combined
 * union of its elements
 *
 * @author Tom Nelson
 */
public class NodeList<B extends Bounded> extends ArrayList<B> implements BoundedList<B>, Bounded {

  private static final Logger log = LoggerFactory.getLogger(NodeList.class);
  private Rectangle2D bounds;

  public NodeList() {}

  public NodeList(int initialCapacity) {
    super(initialCapacity);
  }

  public NodeList(Collection<B> list) {
    for (B node : list) {
      add(node);
    }
  }

  public NodeList(B... nodes) {
    for (B node : nodes) {
      add(node);
    }
  }

  @Override
  public boolean add(B n) {
    if (n instanceof Node) {
      Node node = (Node) n;
      if (node.getParent() == null || !node.getParent().isPresent()) {
        log.error("adding a node {} with unset parent {}", node, node.getParent());
      }
    } else {
      log.error("adding something that is not a Node: {}", n);
    }
    addBoundsFor(n);
    return super.add(n);
  }

  @Override
  public void add(int index, B element) {
    addBoundsFor(element);
    super.add(index, element);
  }

  @Override
  public B remove(int index) {
    B removed = super.remove(index);
    recalculateBounds();
    return removed;
  }

  @Override
  public boolean remove(Object o) {
    boolean removed = super.remove(o);
    recalculateBounds();
    return removed;
  }

  @Override
  public void clear() {
    super.clear();
    bounds = null;
  }

  @Override
  public boolean addAll(Collection<? extends B> c) {
    addBoundsFor(c);
    return super.addAll(c);
  }

  @Override
  public boolean addAll(int index, Collection<? extends B> c) {
    return super.addAll(index, c);
  }

  @Override
  protected void removeRange(int fromIndex, int toIndex) {
    super.removeRange(fromIndex, toIndex);
  }

  @Override
  public boolean removeAll(Collection<?> c) {
    return super.removeAll(c);
  }

  @Override
  public boolean retainAll(Collection<?> c) {
    return super.retainAll(c);
  }

  @Override
  public Rectangle2D getBounds() {
    return bounds;
  }

  private void addBoundsFor(Collection<? extends B> kids) {
    for (B kid : kids) {
      addBoundsFor(kid);
    }
  }

  private void addBoundsFor(B kid) {
    if (bounds == null) {
      bounds = kid.getBounds();
    } else {
      bounds = bounds.createUnion(kid.getBounds());
    }
  }
  /** iterate over all children and update the bounds Called after removing from the collection */
  public void recalculateBounds() {
    bounds = null;
    for (B n : this) {
      addBoundsFor(n);
    }
  }
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/spatial/rtree/NodeList.kt
package edu.uci.ics.jung.visualization.spatial.rtree

import java.awt.geom.Rectangle2D
import org.slf4j.LoggerFactory

/**
 * A list of elements that are Bounded by a Rectangle2D. The bounding box of the list is the union
 * of its elements
 *
 * @author Tom Nelson
 */
open class NodeList<B : Bounded> : ArrayList<B>, BoundedList<B>, Bounded {

  companion object {
    private val log = LoggerFactory.getLogger(NodeList::class.java)
  }

  private var bounds: Rectangle2D? = null

  constructor() : super()

  constructor(initialCapacity: Int) : super(initialCapacity)

  constructor(list: Collection<B>) : super() {
    for (node in list) {
      add(node)
    }
  }

  constructor(vararg nodes: B) : super() {
    for (node in nodes) {
      add(node)
    }
  }

  override fun add(element: B): Boolean {
    if (element is Node<*>) {
      val node = element as Node<*>
      if (node.getParent() == null || !node.getParent().isPresent) {
        log.error("adding a node {} with unset parent {}", node, node.getParent())
      }
    } else {
      log.error("adding something that is not a Node: {}", element)
    }
    addBoundsFor(element)
    return super.add(element)
  }

  override fun add(index: Int, element: B) {
    addBoundsFor(element)
    super.add(index, element)
  }

  override fun removeAt(index: Int): B {
    val removed = super.removeAt(index)
    recalculateBounds()
    return removed
  }

  override fun remove(element: B): Boolean {
    val removed = super.remove(element)
    recalculateBounds()
    return removed
  }

  override fun clear() {
    super.clear()
    bounds = null
  }

  override fun addAll(elements: Collection<B>): Boolean {
    addBoundsForCollection(elements)
    return super.addAll(elements)
  }

  override fun addAll(index: Int, elements: Collection<B>): Boolean {
    return super.addAll(index, elements)
  }

  override fun removeRange(fromIndex: Int, toIndex: Int) {
    super.removeRange(fromIndex, toIndex)
  }

  override fun removeAll(elements: Collection<B>): Boolean {
    return super.removeAll(elements.toSet())
  }

  override fun retainAll(elements: Collection<B>): Boolean {
    return super.retainAll(elements.toSet())
  }

  override fun getBounds(): Rectangle2D {
    return bounds ?: Rectangle2D.Double()
  }

  private fun addBoundsForCollection(kids: Collection<B>) {
    for (kid in kids) {
      addBoundsFor(kid)
    }
  }

  private fun addBoundsFor(kid: B) {
    bounds = bounds?.createUnion(kid.getBounds()) ?: kid.getBounds()
  }

  /** iterate over all children and update the bounds. Called after removing from the collection */
  override fun recalculateBounds() {
    bounds = null
    for (n in this) {
      addBoundsFor(n)
    }
  }
}
